### PR TITLE
created shopping cart and cart item models and added implementations …

### PR DESCRIPTION
**Pull request by irynamatveieva**
…of methods to manage them

### CODE REVIEW

File: src/main/java/com/example/bookstore/controller/ShoppingCartController.java [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+import org.springframework.web.bind.annotation.RestController;
+
+@RequiredArgsConstructor
+@Tag(name = "Shopping cart management", description = "Endpoints for managing shopping cart")
+@RequestMapping(value = "/cart")
+@RestController
+public class ShoppingCartController {
+    private final ShoppingCartService shoppingCartService;
+
+    @GetMapping()
```
Comment (aaananas) [inline src/main/java/com/example/bookstore/controller/ShoppingCartController.java]:
```suggestion
    @GetMapping
```

File: src/main/java/com/example/bookstore/controller/ShoppingCartController.java [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+    @Operation(summary = "Get a shopping cart by user", description = "Get a shopping cart by user")
+    @PreAuthorize("hasRole('ROLE_USER')")
+    public ShoppingCartDto getShoppingCart(Authentication authentication) {
+        User user = (User) authentication.getPrincipal();
+        return shoppingCartService.getShoppingCart(user.getId());
+    }
+
+    @PostMapping()
```
Comment (aaananas) [inline src/main/java/com/example/bookstore/controller/ShoppingCartController.java]:
```suggestion
    @PostMapping
```

File: src/main/java/com/example/bookstore/mapper/ShoppingCartMapper.java [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+import org.mapstruct.Mapping;
+
+@Mapper(config = MapperConfig.class, uses = CartItemMapper.class)
+public interface ShoppingCartMapper {
+    @Mapping(source = "user.id", target = "userId")
+    @Mapping(source = "cartItems", target = "cartItems")
+    ShoppingCartDto toDto(ShoppingCart shoppingCart);
+
+    @Mapping(source = "userId", target = "user.id")
+    @Mapping(source = "cartItems", target = "cartItems")
```
Comment (aaananas) [inline src/main/java/com/example/bookstore/mapper/ShoppingCartMapper.java]: try to keep the same style, in CartItemMapper you used Mappings wrapper annotation

File: src/main/java/com/example/bookstore/model/ShoppingCart.java [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+        if (this == o) {
+            return true;
+        }
+        if (o == null || getClass() != o.getClass()) {
+            return false;
+        }
+        ShoppingCart that = (ShoppingCart) o;
+        return Objects.equals(id, that.id)
+                && Objects.equals(user, that.user)
+                && Objects.equals(cartItems, that.cartItems);
```
Comment (aaananas) [inline src/main/java/com/example/bookstore/model/ShoppingCart.java]:
```suggestion
        return Objects.equals(id, that.id);
```
it ok to use only id here, related entities might not be initialized

**Review by aaananas — CHANGES_REQUESTED** [review]

File: src/main/java/com/example/bookstore/model/ShoppingCart.java [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+    private Long id;
+    @OneToOne
+    @MapsId
+    @JoinColumn(name = "user_id")
+    private User user;
+    @OneToMany(mappedBy = "shoppingCart")
+    private Set<CartItem> cartItems = new HashSet<>();
+
+    @Override
+    public boolean equals(Object o) {
```
Comment (Rommelua) [inline src/main/java/com/example/bookstore/model/ShoppingCart.java]: We **need** equals/hashCode in CartItem class - we have `HashSet<CartItem>`
And we **do not** need  equals/hashCode in ShoppingCart class

File: src/main/java/com/example/bookstore/service/shoppingcart/impl/ShoppingCartServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+    private final CartItemMapper cartItemMapper;
+
+    @Override
+    public ShoppingCartDto getShoppingCart(Long id) {
+        ShoppingCart shoppingCart = getShoppingCartByUserId(id);
+        return shoppingCartMapper.toDto(shoppingCart);
+    }
+
+    @Override
+    public ShoppingCartDto addBookToTheShoppingCart(
```
Comment (Rommelua) [inline src/main/java/com/example/bookstore/service/shoppingcart/impl/ShoppingCartServiceImpl.java]:
```suggestion
    @Override
    @Transactional
    public ShoppingCartDto addBookToTheShoppingCart(
```

File: src/main/java/com/example/bookstore/service/shoppingcart/impl/ShoppingCartServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+            Long id, CreateCartItemRequestDto cartItemRequestDto) {
+        ShoppingCart shoppingCart = getShoppingCartByUserId(id);
+        CartItem cartItem = cartItemMapper.toEntity(cartItemRequestDto);
+        cartItem.setShoppingCart(shoppingCart);
+        cartItemRepository.save(cartItem);
+        Set<CartItem> cartItems = shoppingCart.getCartItems();
+        cartItems.add(cartItem);
+        shoppingCart.setCartItems(cartItems);
```
Comment (Rommelua) [inline src/main/java/com/example/bookstore/service/shoppingcart/impl/ShoppingCartServiceImpl.java]: what is it for?

File: src/main/java/com/example/bookstore/service/shoppingcart/impl/ShoppingCartServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+        return shoppingCartMapper.toDto(getShoppingCartByUserId(id));
+    }
+
+    @Override
+    public CartItemDto updateCartItem(
```
Comment (Rommelua) [inline src/main/java/com/example/bookstore/service/shoppingcart/impl/ShoppingCartServiceImpl.java:46]: @Transactional

File: src/main/java/com/example/bookstore/service/shoppingcart/impl/ShoppingCartServiceImpl.java [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+            Long id, Long cartItemId, UpdateCartItemRequestDto cartItemRequestDto) {
+        CartItem cartItem = cartItemRepository.findById(cartItemId).orElseThrow(
+                () -> new RuntimeException("Can`t find cartItem by id" + id)
+        );
+        cartItem.setQuantity(cartItemRequestDto.getQuantity());
+        CartItem savedCartItem = cartItemRepository.save(cartItem);
+        return cartItemMapper.toDto(savedCartItem);
+    }
+
+    @Override
```
Comment (Rommelua) [inline src/main/java/com/example/bookstore/service/shoppingcart/impl/ShoppingCartServiceImpl.java:56]: @Transactional

File: src/main/java/com/example/bookstore/service/shoppingcart/impl/ShoppingCartServiceImpl.java [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+    public void removeCartItem(Long id, Long cartItemId) {
+        ShoppingCart shoppingCart = getShoppingCartByUserId(id);
+        CartItem cartItem = cartItemRepository.findById(cartItemId).orElseThrow(
+                () -> new RuntimeException("Can`t find cartItem by id" + id)
+        );
+        Set<CartItem> cartItems = shoppingCart.getCartItems();
+        cartItems.remove(cartItem);
+        cartItemRepository.delete(cartItem);
+        shoppingCart.setCartItems(cartItems);
```
Comment (Rommelua) [inline src/main/java/com/example/bookstore/service/shoppingcart/impl/ShoppingCartServiceImpl.java]:
```suggestion
        cartItemRepository.delete(cartItem);
```

**Review by Rommelua — CHANGES_REQUESTED** [review]

**Pull request merged**